### PR TITLE
run integration test on ci

### PR DESCRIPTION
**Pull request by andreykiryushkin**
There are testing instructions in `TESTING.md` which need to be executed manually.
This PR introduces an automatic integration test suite based on these rules which runs on each push.

### CODE REVIEW

File: .github/workflows/integration-testing.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      - '**/docs/**/*'
+
+jobs:
+  integration-test:
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 15
+
+    strategy:
+      matrix:
+        node-version: [16]
```
Comment (tolbrino) [inline .github/workflows/integration-testing.yaml:22]: should be 14

File: scripts/integration-test.sh [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+# prevent souring of this script, only allow execution
+$(return >/dev/null 2>&1)
+test "$?" -eq "0" && { echo "This script should only be executed." >&2; exit 1; }
+          
+# set log id and use shared log function for readable logs
+declare mydir
+mydir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &>/dev/null && pwd -P)
+declare HOPR_LOG_ID="hopr-connect-test"
+source "${mydir}/../scripts/utils.sh"
```
Comment (tolbrino) [inline scripts/integration-test.sh:11]:
```suggestion
source "${mydir}/utils.sh"
```

File: scripts/integration-test.sh [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+declare server_pid
+
+declare client0_log="${tmp}/hopr-connect-client0.log"
+declare client0_pid
+
+declare client1_log="${tmp}/hopr-connect-client1.log"
+declare client1_pid
+
+function free_ports {
+    for port in 9090 9091 9092; do
```
Comment (tolbrino) [inline scripts/integration-test.sh:34]: better use ports which don't conflict with our standard ports in testing

File: scripts/integration-test.sh [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+}
+trap cleanup SIGINT SIGTERM ERR EXIT
+
+function start_node() {
+    declare filename=${1}
+    declare log_file=${2}
+    declare rest_args=${@:3}
+
+    DEBUG=hopr-connect*,simple-peer \
+    yarn dlx \
```
Comment (tolbrino) [inline scripts/integration-test.sh:65]: why all the separate lines?

File: scripts/utils.sh [hunk 5]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env bash
+
+# prevent execution of this script, only allow execution
```
Comment (tolbrino) [inline scripts/utils.sh:3]:
```suggestion
# prevent execution of this script, only allow sourcing
```

File: scripts/utils.sh [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+$(return >/dev/null 2>&1)
+test "$?" -eq "0" || { echo "This script should only be sourced." >&2; exit 1; }
+
+# exit on errors, undefined variables, ensure errors in pipes are not hidden
+set -Eeuo pipefail
+
+# $1=version string, semver
+function get_version_maj_min() {
```
Comment (tolbrino) [inline scripts/utils.sh:11]: not needed

File: scripts/utils.sh [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  echo $(get_version_maj_min_pat $1 | cut -d. -f1,2)
+}
+
+# $1=version string, semver
+function get_version_maj_min_pat() {
```
Comment (tolbrino) [inline scripts/utils.sh:16]: not needed

Comment (tolbrino) [reply scripts/utils.sh:16]: please just remove all functions which are not needed

**Review by tolbrino — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-testing.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
+name: HOPR Connect integration tests
+
+env:
+  HOPR_GITHUB_REF: ${{ github.ref }}
+  MATRIX_ROOM: ${{ secrets.MATRIX_ROOM }}
```
Comment (tolbrino) [inline .github/workflows/integration-testing.yaml:5]: remove env vars which are not used

File: scripts/integration-test.sh [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+set -Eeuo pipefail
+
+usage() {
+  msg
+  msg "Usage: $0"
+  msg
+}
+
+# setup paths
+declare tmp="/tmp"
```
Comment (tolbrino) [inline scripts/integration-test.sh:23]: this will fail on macos, since there it must be /var/tmp

File: scripts/integration-test.sh [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  trap - SIGINT SIGTERM ERR EXIT
+
+  # Cleaning up everything
+  log "Stopping up processes"
+  for pid in ${server_pid} ${client0_pid} ${client1_pid}; do
+    kill -9 ${pid}
+    log "killed ${pid}"    
+  done
+
+  free_ports
```
Comment (tolbrino) [inline scripts/integration-test.sh:53]: why duplicate the effort you've done above? one should be enough

File: scripts/integration-test.sh (hunk 4 above)
Comment (andreykiryushkin) [reply scripts/integration-test.sh:58]: Readability fan

File: scripts/integration-test.sh (hunk 10 above)
Comment (andreykiryushkin) [reply scripts/integration-test.sh:46]: Not sure what you mean. I kill the processes occupying ports on startup and in the exit trap, that's not duplication

File: .github/workflows/integration-testing.yaml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            node_modules
+            */*/node_modules
+            ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-nodejs-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-nodejs-
+
+      - name: Build
+        run: |
+          yarn set version berry
```
Comment (robertkiel) [inline .github/workflows/integration-testing.yaml:47]: Current versions of yarn should interprete `.yarnrc.yml` and automatically switch to yarn 2

**Review by robertkiel — APPROVED** [review]
`examples/client.ts` and `examples/server.ts` might need some refactor but this can be done later

**Review by tolbrino — APPROVED** [review]

**Pull request merged**